### ISSUE TITLE
Overload error for kdc101.py

### ISSUE THREAD

**Issue by pya** [MSLNZ/msl-equipment#1]
I tried to run `msl\examples\equipment\resources\kdc101.py` to control an MT550/M-Z8 with KDC101  but always get an *Overload* error on the display of the hardware box. I need to reboot after this. Furthermore, the sound when moving is much different than when suing the Kenesis GUI. The pitch is much higher.

This simplified script reproduces the error:

```
import os
import time

from msl.equipment import EquipmentRecord, ConnectionRecord, Backend
from msl.equipment.resources.thorlabs.kinesis.enums import UnitType

from msl.equipment import EquipmentRecord, ConnectionRecord, Backend

# you must update the following values
kinesis_path = 'C:/Program Files/Thorlabs/Kinesis'
serial_number = '27002488'

# the Thorlabs.MotionControl.KCube.DCServo.dll depends on other DLLs from Thorlabs
# make sure to add the Kinesis folder to the environment PATH
os.environ['PATH'] += os.pathsep + kinesis_path

record = EquipmentRecord(
    manufacturer='Thorlabs',
    model='KDC101',
    serial=serial_number,
    connection=ConnectionRecord(
        backend=Backend.MSL,
        address='SDK::KCubeDCServo::Thorlabs.MotionControl.KCube.DCServo.dll',
    ),
)

motor = record.connect()
print(motor)

motor.identify()

time.sleep(5)

motor.start_polling(250)

info = motor.get_hardware_info()
print('Found device: {}'.format(info.notes))

before = motor.get_move_relative_distance()
new = 2
motor.set_move_relative_distance(new)
time.sleep(1)
after = motor.get_move_relative_distance()

print(f'Old distance {before}, desired, {new}, got {after}')

print('Start position', motor.get_position())
motor.get_position()

time.sleep(1)

motor.move_relative_distance()

time.sleep(1)


print('End position', motor.get_position())
motor.get_position()

time.sleep(1)

motor.stop_polling()
motor.close()

```
 produces this kind of output:

```

Failed to open C:\ProgramData\Thorlabs\MotionControl\ThorlabsCustomSettings.xml, error id=2, No such file or directory
XMLDocument error id=3 str1=C:\ProgramData\Thor str2=
Failed to open C:\ProgramData\Thorlabs\MotionControl\ThorlabsSettingsOverrides.xml, error id=2, No such file or directory
XMLDocument error id=3 str1=C:\ProgramData\Thor str2=
Failed to open C:\ProgramData\Thorlabs\MotionControl\ThorlabsCustomSettings.xml, error id=2, No such file or directory
XMLDocument error id=3 str1=C:\ProgramData\Thor str2=
Failed to open C:\ProgramData\Thorlabs\MotionControl\ThorlabsSettingsOverrides.xml, error id=2, No such file or directory
XMLDocument error id=3 str1=C:\ProgramData\Thor str2=
KCubeDCServo<Thorlabs|KDC101|27002488>
Found device: b'KDC101 DC Servo Drive'
Old distance 2, desired, 2, got 2
Start position 0
End position -1212
````

The number of the end position is always different. The use of other move functions such as `move_to_position`  result in the same error.

I am on Windows 10 Home 64 bit with the 64 bit DLLs and use Python 3.6
`'3.6.5 | packaged by conda-forge | (default, Apr  6 2018, 16:13:55) [MSC v.1900 64 bit (AMD64)]'`.

**Comment by pya** [comment]
One exponential cause could be a wrong data type for a one or more parameters when calling a DLL function.  I played around quite a bit with `msl\equipment\resources\thorlabs\kinesis\api_functions.py` but always got an error. It seemed different for different values for `c_int` (`c_int32`, `c_int64`, `c_int16`) for `CC_SetMoveRelativeDistance`, causing the sled to move a lot further. But there was a lot of variation in the reaction with the same data type for different runs.

**Comment by jborbely** [comment]
What is the version number of Kinesis that you are using? I will test the script against that version when I get to my lab in the morning.

**Comment by pya** [comment]
I am using the DLLS from this version: 1.14.8.0 64 bit.

**Comment by jborbely** [comment]
There is something strange happening with Kinesis v1.14.8. The problem seems to be when the program calls the `CC_LoadSettings(char const * serialNo);` function (which happens by default in the constructor of the `KCubeDCServo` class). Apparently, this is not an issue in v1.11.2 (which is the version that I am using). 

As a temporary solution you can include **properties={'load_settings': False}** in the ``ConnectionRecord``, i.e.:
```python
    record = EquipmentRecord(
        manufacturer='Thorlabs',
        model='KDC101',
        serial=serial_number,
        connection=ConnectionRecord(
            backend=Backend.MSL,
            address='SDK::KCubeDCServo::Thorlabs.MotionControl.KCube.DCServo.dll',
            properties={'load_settings': False}
        ),
    )
```
to not automatically load the settings.

The reason for calling `CC_LoadSettings` is because `CC_GetDeviceUnitFromRealValue` and `CC_GetRealValueFromDeviceUnit` require knowledge of the stage to perform the calculation. This can either be done in one of two ways:

(1) If the software has been configured using Kinesis then calling `CC_LoadSettings` will load all the settings for the associated stage.
(2) Alternatively using `CC_SetMotorParamsExt` will define the specific parameters. The required values can be read from the *ThorlabsDefaultSettings.xml* file (for me it is located in *C:\Program Files\Thorlabs\Kinesis*). You will need to search for the **Pitch**, **StepsPerRev** and **GearboxRatio** for your specific stage. For example, I have a Z825B connected to my KDC101 and from the .xml file this stage has the values:
```xml
<DeviceSettingsDefinition Name="Z825">
    ...
        <Pitch>1.0</Pitch>
        <StepsPerRev>512</StepsPerRev>
        <GearboxRatio>67</GearboxRatio>
    ...
```
Until either (1) or (2) are successful you cannot use the `get_real_value_from_device_unit` nor the `get_device_unit_from_real_value` functions and you must work in Device Units. 

I will have to do some more tests to find out what is causing the issue in v1.14.8, especially since the `CC_LoadSettings` function only takes the serial number as an input and I configured Kinesis to know that the Z825B stage is connected to it.

**Comment by pya** [comment]
Thanks. I will try this out as soon as I have access to the hardware system in a day or two. I am using a MTS50-Z8 motor. The XML file has the same values for `Pitch`, `StepsPerRev`, and `GearboxRatio` as you list above.

**Comment by pya** [comment]
It works with `properties={'load_settings': False}` and sticking to device units only.

But setting the motor parameters manually:

``` 
motor = record.connect()
motor.set_motor_params(steps_per_rev=512, gear_box_ratio=67, pitch=1.0)
```

leads to this exception: 

```
(motor) C:\Users\mike\motor\tests>python kdc101.py
Failed to open C:\ProgramData\Thorlabs\MotionControl\ThorlabsCustomSettings.xml, error id=2, No such file or directory
XMLDocument error id=3 str1=C:\ProgramData\Thor str2=
Failed to open C:\ProgramData\Thorlabs\MotionControl\ThorlabsSettingsOverrides.xml, error id=2, No such file or directory
XMLDocument error id=3 str1=C:\ProgramData\Thor str2=
Failed to open C:\ProgramData\Thorlabs\MotionControl\ThorlabsCustomSettings.xml, error id=2, No such file or directory
XMLDocument error id=3 str1=C:\ProgramData\Thor str2=
KCubeDCServo<Thorlabs|KDC101|27002488 at SDK::KCubeDCServo::Thorlabs.MotionControl.KCube.DCServo.dll> FT_InvalidHandle: The FTDI functions have not been initialized
Traceback (most recent call last):
  File "kdc101.py", line 36, in <module>
    motor.set_motor_params(steps_per_rev=512, gear_box_ratio=67, pitch=1.0)
  File "C:\Users\mike\Miniconda3\envs\motor\lib\site-packages\msl\equipment\resources\thorlabs\kinesis\kcube_dc_servo.py", line 1734, in set_motor_params
    self.set_motor_params_ext(steps_per_rev, gear_box_ratio, pitch)
  File "C:\Users\mike\Miniconda3\envs\motor\lib\site-packages\msl\equipment\resources\thorlabs\kinesis\kcube_dc_servo.py", line 1760, in set_motor_params_ext
    self.sdk.CC_SetMotorParamsExt(self._serial, steps_per_rev, gear_box_ratio, pitch)
  File "C:\Users\mike\Miniconda3\envs\motor\lib\site-packages\msl\equipment\resources\thorlabs\kinesis\motion_control.py", line 116, in errcheck_api
    self.raise_exception('{}: {}'.format(*ERROR_CODES[result]))
  File "C:\Users\mike\Miniconda3\envs\motor\lib\site-packages\msl\equipment\connection.py", line 85, in raise_exception
    raise self._exception_handler('{!r}\n{}'.format(self, msg))
msl.equipment.exceptions.ThorlabsError: KCubeDCServo<Thorlabs|KDC101|27002488 at SDK::KCubeDCServo::Thorlabs.MotionControl.KCube.DCServo.dll>
FT_InvalidHandle: The FTDI functions have not been initialized

```

I worked round it by manually converting mm into devise units and back:

```
class UnitConverter:

    def __init__(self, units_per_mm=34304):
        self.units_per_mm = units_per_mm

    def to_mm(self, index):
        return index / self.units_per_mm

    def to_device_units(self, mm):
        return int(mm * self.units_per_mm)
```

**Comment by jborbely** [comment]
Thanks for the feedback. I get the same issue in a C++ program and also using Kinesis v1.14.9. I have contacted Thorlabs.

**Comment by jborbely** [comment]
The summary from my email exchanges with a Thorlabs representative is as follows:

1. The functions `CC_LoadSettings` and `CC_LoadNamedSettings` have known bugs. Do not use them.
2. Either the documentation for the returned value or the returned value itself for the `CC_SetMotorParamsExt` function is wrong. If the function returns a value of 1 then according to the documentation that means that an error occurred (the `FT_InvalidHandle` error). However, if you don't do error checking on the returned value and instead call `CC_GetMotorParamsExt` to test if the motor parameters were successfully set then you will discover that the motor parameters were set successfully. The Thorlabs representative did not comment as to whether it was a documentation error (i.e., 1 means success) or if the returned value is wrong.
3. There is no indication as to whether these bugs will be fixed in the next release nor when the next release date will be.

I will keep this issue open until I implement my own solution to these known Kinesis bugs. But, fundamentally, these issues are a Kinesis issue and not an issue that results from my code.

**Comment by pya** [comment]
Thanks. Looks like we live with the workaround for a while.

**Comment by jborbely** [comment]
Thorlabs has recently released v1.14.10 of Kinesis. In this version the ``CC_LoadSettings`` function appears to be working properly. 

The [kdc101.py](https://github.com/MSLNZ/msl-equipment/blob/master/msl/examples/equipment/resources/thorlabs/kdc101.py) example works for me with Kinesis v1.14.10.

Please be aware that I have recently made a rather large change to the API and so you may need to re-install the repository if the example does not run for you. Feel free to open a new issue if you have any problems updating your code to reflect the API changes.

_NOTE: I have tested other ``XXX_LoadSettings`` functions for other controllers and it does not appear to be fixed for all controllers, but for KDC101 it seems okay._

**Comment by pya** [comment]
Thanks for taking car of this. Currently, I don't have access to the hardware, but I will ping the person who does.
